### ISSUE TITLE
Prediction with expanded x axis is too confident?

### ISSUE THREAD

**Issue by jeffwillette** [EmilienDupont/neural-processes#1]
This is not really an issue with the code, but something that I expected to work differently with neural processes and I would value your opinion. 

When making a prediction, if I expand the x axis to a wider range I would expect the functions to become less certain and more varied. When I tried expanding the axis this is what I actually saw happen...

![np-prediction](https://user-images.githubusercontent.com/9424192/64840199-50e05280-d635-11e9-95de-0f5d0d599dea.png)

I am having a hard time justifying why this would be the case that the neural process remains confident in its prediction and every function just goes off in the same direction. Do you have any insight?

**Comment by EmilienDupont** [comment]
Hi, this is a good question and something I was also wondering about when first implementing neural processes. 

The neural process in the above figure was trained on data in the [-pi, pi] range. When expanding the x-axis you are therefore expanding outside the range of the training data. Unlike Gaussian processes, the behavior of neural processes is basically "undefined" far from the training data. It could be uncertain or it could be confident, but this is not something we have control over. In practice though, I have mostly seen the models collapsing to very confident predictions outside of the training data (other people I have spoken to about this seem to observe the same behavior).

Of course, the neural process should be able to capture uncertainty within the training data range. For example, if you looked at the [-pi, pi] range and fixed just one context point, then the neural process samples should have less uncertainty around the fixed context point and more uncertainty far away from it.

Anyway, this is all based on my experience and discussions with other people, it might be worth emailing the authors of the paper too for more details on this 😁

**Comment by jeffwillette** [comment]
I have an email out to the first author. Ill post here if and when I get a response, thanks for your input!

**Comment by jeffwillette** [comment]
I noticed that when I visualize the untrained prior on a large range, it tends to drift off in a downward direction on both sides just like in the image above which might cause that downward bias. Shouldn't a randomly initialized prior over functions be going off in all directions?

I'm trying to play around with different initialization techniques to see if it helps. This blog here seems to think that it is based on the initialization https://chrisorm.github.io/NGP.html

**Comment by EmilienDupont** [comment]
It is kind of surprising that the untrained prior also tends to drift off in a downward direction. However, it is not necessarily the case that a randomly initialized prior will produce functions that go off in all directions. This also depends on specifics of the model architecture and activation functions used. Kaspar's blog on Neural Processes has some good examples of this: https://kasparmartens.rbind.io/post/np/

For example, if you change the activation function from a sigmoid to a ReLU, the draws from the prior will look different. I'm not sure specifically what aspect of the architecture makes it drift off in a downward direction for large values of x in this case, but it would be interesting to look into.

**Comment by jeffwillette** [comment]
Actually, after I sampled it more times (from more random initializations of the model) I found that it doesn't always go down, but they do head of in "clumps" in one direction or another (sometimes up) from the 0 point. I can't explain why this happens, but here is an example, that happens to go down in this case)

![biased-prior](https://user-images.githubusercontent.com/9424192/65853589-f9f8ae00-e394-11e9-9c43-3d08dca7d23b.png)

from my experimenting I have found that unless I normalize the x axis values around 0, I get terrible predictions. Maybe this is from the biased prior? I was able to get a less bias in the prior by using orthogonal initialization pictured below, but I eventually settled on just having to normalize everything around the 0 range which is probably a good thing to do anyway. I had never thought about switching the activation functions. I will read the blog post you sent. Thanks for your input 

![prior-orthogonal-init](https://user-images.githubusercontent.com/9424192/65853607-0bda5100-e395-11e9-8228-2342b6e99af5.png)
